plugins/auto-sizes/hooks.php

```diff
@@ -139,79 +139,139 @@ function auto_sizes_get_width( string $layout_width, int $image_width ): string
+			continue;
 		}
 
-		// Resize image width.
-		if ( isset( $parsed_block['attrs']['width'] ) ) {
-			$processor->set_attribute( 'data-resize-width', $parsed_block['attrs']['width'] );
+		if ( preg_match( '/wp-image-([0-9]+)/i', $class, $class_id ) === 1 ) {
+			$attachment_id = absint( $class_id[1] );
+			if ( $attachment_id > 0 ) {
+				$images[] = $attachment_id;
+			}
 		}
+	}
+
+	// Reduce the array to unique attachment IDs.
+	$attachment_ids = array_unique( array_filter( $images ) );
 
-		$content = $processor->get_updated_html();
+	if ( count( $attachment_ids ) > 1 ) {
+		/*
+		 * Warm the object cache with post and meta information for all found
+		 * images to avoid making individual database calls.
+		 */
+		_prime_post_caches( $attachment_ids, false, true );
 	}
+
 	return $content;
 }
-add_filter( 'render_block_core/image', 'auto_sizes_filter_image_tag', 10, 2 );
-add_filter( 'render_block_core/cover', 'auto_sizes_filter_image_tag', 10, 2 );
+add_filter( 'the_content', 'auto_sizes_prime_attachment_caches', 6 );
 
 /**
  * Filter the sizes attribute for images to improve the default calculation.
  *
  * @since 1.1.0
  *
- * @param string $content The block content about to be rendered.
+ * @param string                                                   $content      The block content about to be rendered.
+ * @param array{ attrs?: array{ align?: string, width?: string } } $parsed_block The parsed block.
+ * @param WP_Block                                                 $block        Block instance.
  * @return string The updated block content.
  */
-function auto_sizes_improve_image_sizes_attributes( string $content ): string {
+function auto_sizes_filter_image_tag( string $content, array $parsed_block, WP_Block $block ): string {
 	$processor = new WP_HTML_Tag_Processor( $content );
-	if ( ! $processor->next_tag( array( 'tag_name' => 'img' ) ) ) {
-		return $content;
-	}
+	$has_image = $processor->next_tag( array( 'tag_name' => 'IMG' ) );
 
-	$remove_data_attributes = static function () use ( $processor ): void {
-		$processor->remove_attribute( 'data-needs-sizes-update' );
-		$processor->remove_attribute( 'data-align' );
-		$processor->remove_attribute( 'data-resize-width' );
-	};
+	// Only update the markup if an image is found.
+	if ( $has_image ) {
+
+		/**
+		 * Callback for calculating image sizes attribute value for an image block.
+		 *
+		 * This is a workaround to use block context data when calculating the img sizes attribute.
+		 *
+		 * @param string $sizes The image sizes attribute value.
+		 * @param string $size  The image size data.
+		 */
+		$filter = static function ( $sizes, $size ) use ( $block ) {
+			$id        = $block->attributes['id'] ?? 0;
+			$alignment = $block->attributes['align'] ?? '';
+			$width     = $block->attributes['width'] ?? '';
+
+			// Hypothetical function to calculate better sizes.
+			$sizes = auto_sizes_calculate_better_sizes( $id, $size, $alignment, $width );
```

Comment:
Perhaps overkill, but we don't know for sure what types that `$block->attributes` contains.

```suggestion
			$sizes = auto_sizes_calculate_better_sizes( (int) $id, (string) $size, $alignment, (string) $width );
```